File: speech_to_text/csharp/async/remote_file/RemoteFile.cs
using System;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading.Tasks;

// Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
string? apiKey = Environment.GetEnvironmentVariable("SONIOX_API_KEY");
if (string.IsNullOrEmpty(apiKey))
{
    throw new InvalidOperationException("SONIOX_API_KEY is not set.");
}
string apiBase = "https://api.soniox.com";
string audioUrl = "https://soniox.com/media/examples/coffee_shop.mp3";

var client = new HttpClient();
client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);

async Task pollUntilComplete(string transcriptionId) {
    while (true)
    {
        var response = await client.GetAsync($"{apiBase}/v1/transcriptions/{transcriptionId}");
        response.EnsureSuccessStatusCode();
        var responseJson = await response.Content.ReadAsStringAsync();
        var transcription = JsonNode.Parse(responseJson);
        var status = transcription?["status"]?.ToString();
        if (string.IsNullOrEmpty(status))
        {
            throw new InvalidOperationException("Status is missing in transcription response.");
        }
        if (status == "completed")
        {
            return;
        }
        else if (status == "error")
        {
            var errorMessage = transcription?["error_message"]?.ToString() ?? "Unknown error";
            throw new Exception($"Transcription error: {errorMessage}");
        }
        await Task.Delay(1000);
    }
}

Console.WriteLine("Starting transcription...");

var request = new StringContent(
    JsonSerializer.Serialize(new {
        audio_url = audioUrl,
        model = "stt-async-preview",
        language_hints = new []{"en", "es"}
    }),
    System.Text.Encoding.UTF8,
    "application/json"
);
var response = await client.PostAsync($"{apiBase}/v1/transcriptions", request);
response.EnsureSuccessStatusCode();
var responseJson = await response.Content.ReadAsStringAsync();
var transcription = JsonNode.Parse(responseJson);
var transcriptionId = transcription?["id"]?.ToString();
if (string.IsNullOrEmpty(transcriptionId))
{
    throw new InvalidOperationException("Transcription ID is missing.");
}

Console.WriteLine($"Transcription ID: {transcriptionId}");

await pollUntilComplete(transcriptionId);

// Get the transcript text
response = await client.GetAsync($"{apiBase}/v1/transcriptions/{transcriptionId}/transcript");
response.EnsureSuccessStatusCode();
responseJson = await response.Content.ReadAsStringAsync();
var transcript = JsonNode.Parse(responseJson);
var transcriptText = transcript?["text"]?.ToString();
if (string.IsNullOrEmpty(transcriptText))
{
    throw new InvalidOperationException("Transcript text is missing.");
}
Console.WriteLine("Transcript:");
Console.WriteLine(transcriptText);

File: speech_to_text/go/async/remote_file/remote_file.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"time"
)

// Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
var apiKey = os.Getenv("SONIOX_API_KEY")
var apiBase = "https://api.soniox.com"
var audioUrl = "https://soniox.com/media/examples/coffee_shop.mp3"

func pollUntilComplete(transcriptionId string) {
	for {
		req, err := http.NewRequest("GET", fmt.Sprintf("%s/v1/transcriptions/%s", apiBase, transcriptionId), nil)
		if err != nil {
			panic(err)
		}
		req.Header.Set("Authorization", "Bearer "+apiKey)
		res, err := http.DefaultClient.Do(req)
		if err != nil {
			panic(err)
		}
		defer res.Body.Close()
		if res.StatusCode != http.StatusOK {
			panic(res.Status)
		}
		var statusResponse map[string]interface{}
		if err := json.NewDecoder(res.Body).Decode(&statusResponse); err != nil {
			panic(err)
		}
		if status, ok := statusResponse["status"].(string); ok {
			if status == "error" {
				panic("Transcription error: " + statusResponse["error_message"].(string))
			} else if status == "completed" {
				break
			}
		}
		time.Sleep(1 * time.Second)
	}
}

func main() {
	fmt.Println("Starting transcription...")

	transcriptionPayload := map[string]any{
		"audio_url":      audioUrl,
		"model":          "stt-async-preview",
		"language_hints": []string{"en", "es"},
	}
	transcriptionBody, err := json.Marshal(transcriptionPayload)
	if err != nil {
		panic(err)
	}
	req, err := http.NewRequest("POST", fmt.Sprintf("%s/v1/transcriptions", apiBase), bytes.NewReader(transcriptionBody))
	if err != nil {
		panic(err)
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Content-Type", "application/json")
	res, err := http.DefaultClient.Do(req)
	if err != nil {
		panic(err)
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusCreated {
		panic(res.Status)
	}
	var transcription map[string]interface{}
	if err := json.NewDecoder(res.Body).Decode(&transcription); err != nil {
		panic(err)
	}
	transcriptionId := transcription["id"].(string)

	fmt.Println("Transcription ID: " + transcriptionId)

	pollUntilComplete(transcriptionId)

	// Get the transcript text
	req, err = http.NewRequest("GET", fmt.Sprintf("%s/v1/transcriptions/%s/transcript", apiBase, transcriptionId), nil)
	if err != nil {
		panic(err)
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)
	res, err = http.DefaultClient.Do(req)
	if err != nil {
		panic(err)
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusOK {
		panic(res.Status)
	}
	var transcriptResponse map[string]interface{}
	if err := json.NewDecoder(res.Body).Decode(&transcriptResponse); err != nil {
		panic(err)
	}
	fmt.Printf("Transcript:\n%s\n", transcriptResponse["text"].(string))

	// Delete the transcription
	req, err = http.NewRequest("DELETE", fmt.Sprintf("%s/v1/transcriptions/%s", apiBase, transcriptionId), nil)
	if err != nil {
		panic(err)
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)
	res, err = http.DefaultClient.Do(req)
	if err != nil {
		panic(err)
	}
	defer res.Body.Close()
	if res.StatusCode != http.StatusNoContent {
		panic(res.Status)
	}
}
